src/keyutils_persistent.rs

```diff
@@ -0,0 +1,140 @@
+/*!
+
+# keyutils-persistent credential store
+
+TODO
+
+ */
+use std::collections::HashMap;
+
+use crate::credential::{
+    Credential, CredentialApi, CredentialBuilder, CredentialBuilderApi, CredentialPersistence,
+};
+use crate::keyutils::KeyutilsCredential;
+use crate::secret_service::SsCredential;
+use crate::{Error, Result};
+
+#[derive(Debug, Clone)]
+pub struct KeyutilsPersistentCredential {
+    keyutils: KeyutilsCredential,
+    ss: SsCredential,
+}
+
+impl CredentialApi for KeyutilsPersistentCredential {
+    fn set_password(&self, password: &str) -> Result<()> {
+        self.set_secret(password.as_bytes())
+    }
+
+    fn set_secret(&self, secret: &[u8]) -> Result<()> {
+        let prev_secret = self.keyutils.get_secret()?;
+        self.keyutils.set_secret(secret)?;
+
+        if let Err(err) = self.ss.set_secret(secret) {
+            self.keyutils.set_secret(&prev_secret)?;
+            return Err(err);
+        }
+
+        Ok(())
+    }
+
+    fn get_password(&self) -> Result<String> {
+        if let Ok(password) = self.keyutils.get_password() {
+            return Ok(password);
+        }
+
+        let password = self.ss.get_password().map_err(ambigous_to_no_entry)?;
+        self.keyutils.set_password(&password)?;
+
+        Ok(password)
+    }
+
+    fn get_secret(&self) -> Result<Vec<u8>> {
+        if let Ok(secret) = self.keyutils.get_secret() {
+            return Ok(secret);
+        }
+
+        let secret = self.ss.get_secret().map_err(ambigous_to_no_entry)?;
+        self.keyutils.set_secret(&secret)?;
+
+        Ok(secret)
+    }
+
+    fn get_attributes(&self) -> Result<HashMap<String, String>> {
+        let Ok(attributes) = self.ss.get_attributes() else {
+            return self.keyutils.get_attributes();
```

Comment:
Now that we took the direction of the persistent keyutils rather than secret service with cache, I would say that it's not necessary to implement attributes-related function (like for the keyutils credential). Let me remove them.